### ISSUE TITLE
User notifications on invoice transition in finance flow

### ISSUE THREAD

**Issue by sks444** [HyphaApp/hypha#2704]
## Summary

OTF needs to inform the users involved in the process of invoice transition through the webapp.  This is so they can be kept up-to-date of the invoice status, and be notified if they are required to do anything.

## Users involved
- Contractor: they submit the invoice.

- Programme Manager Staff: they are the contractor's point of contact within OTF.  They review the invoice to make sure the deliverables have been met correctly. They review the invoice, communicate with the contractor if changes are made.

- Finance 1: they are the finance staff that review the invoice to check it has the required information and all numbers are correct.

- Finance 2: they are the finance staff that review the invoice, cross-check the work carried out by the finance 1 user. They also send the invoice to the payment system if all looks OK.   

## Notification method

It's been decided that 1) OTF Staff will receive notifications via Slack, and 2) contractors will receive notifications via the email address defined in their account. 

## Communication style

The current OTF-to-contractor communication style is to provide the minimum level of detail in the email and to require them to log-in to read the full message.

## Invoice states and who needs to be notified

The first sequence diagram below outlines the possible invoice states (states 1-17 below).

It also defines which user needs to be notified at each change of status (Yellow boxes a-to-s below).

![image](https://user-images.githubusercontent.com/1899506/153173085-045f0503-3f18-4252-843c-63fdb68d78ec.png)
(If it's hard to read, you [can see the original](https://notes.saneux.ei8fdb.org/shares/pI92hWbkbGPe09lysF9fWH))

### If an invoice is rejected
These sequence diagram outline if an invoice is rejected.

#### PM declines invoice
![image](https://user-images.githubusercontent.com/1899506/153172391-a6b5314b-1d47-4141-b6f6-00af8c93c4c7.png)

#### Finance 1 declines invoice
![image](https://user-images.githubusercontent.com/1899506/159530135-a9f6a8d8-071d-4993-8e73-2342a1b0f9b8.png)

#### Finance 2 declines invoice
![image](https://user-images.githubusercontent.com/1899506/159530718-c4ff118b-dc4e-4ff7-819e-2f6c8035866b.png)

## Notification message content
Draft [notification messages are ready](https://miro.com/app/board/uXjVOO-j-PQ=/?moveToWidget=3458764518522959381&cot=14) for review.

## Feedback
- [x] get feedback on notification message content from @fourthletter 
- [x] get feedback from @frjo, @sks444 focused on difficulties to implementation 
- [ ] get feedback on notification message content from users (contractors, PMs, finance)
- [ ] get final agreement/sign-off

## Open questions
- [ ] For Finance how long can it take for a payment to leave the OTF banking system and reach a contractors account? (Take into account the different global locations bank accounts can be. Some will take longer or short periods of time)
- [ ] For finance & PMs: If Finance 2 declines an invoice should the PM be notified?

**Comment by sks444** [comment]
We need a way to notify(either via slack or email) to respective users whenever an invoice is getting transitioned into different states. Above are the few notifications that I collected from different finance issues. We can discuss here and decide something final. I believe we also need to send notification to applicant on final approval of invoice(finance 2 approval)?

For now, I will go-ahead and complete the finance flow in hypha without these notifications. And come back to this afterwards.

cc: @frjo @ei8fdb @fourthletter

**Comment by fourthletter** [comment]
On Feb 3 during standup we discussed the content that should be embedded into emails. 

- [ ] Default settings for emails and slack; different porks for different forks: staff gets slack while contractors get emails
- [ ] Emails will contain links that go directly to the projects page, invoice, communication/message, etc. 
- [ ] Bernard has ideas for additional email notifications
- [ ] Fredrik propose we keep similar email templates used in Submissions
- [ ] OTF creates new slack channel for project related notifications (not directly pertinent to this GH issue)
- [ ] Messages/Notifications the SAME for Slack?
- [ ] Fredrik is probably already thinking about this, but we may need to build something in the WagTail admin.

### How does staff receive slack notifications for specific funds/labs? Image of the lab/fund form:
<img width="839" alt="Screen Shot 2022-02-03 at 1 01 34 PM" src="https://user-images.githubusercontent.com/20019656/152339248-61c3ef31-c751-48b7-b23d-859fbd69214a.png">

### How does each staff member sign up for slack notifications generally? Image of the profile page:

<img width="589" alt="Screen Shot 2022-02-03 at 1 05 16 PM" src="https://user-images.githubusercontent.com/20019656/152339599-2f67f42b-4ca0-4cee-ba53-cc3ef1e721f5.png">

**Comment by ei8fdb** [comment]
@fourthletter going back through my notes, these are the different states and review steps spoken about:

![image](https://user-images.githubusercontent.com/1899506/152383410-279cfd1a-b4a4-4162-812b-03d32c04b447.png)

Can you confirm th following are correct:
- when finance 1 requests changes the request goes from finance 1 to PM  staff (5 above), then from PM Staff to contractor (6 above), back to PM staff (7 above), then PM staff to finance 1 (8 above) 
- when finance 2 requests changes the request goes from finance 2 to PM Staff (10 above) , then from PM staff to contractor (11 above), back to PM staff (12 above), then from PM Staff to finance 1 (13 above), then from finance 1 to finance 2 (14 above).

There's a few more steps than @sks444 mentioned above. Once you can tell us that, then we can write the notification content.

**Comment by fourthletter** [comment]
@ei8fdb 

These are correct:
- [x] when finance 1 requests changes the request goes from finance 1 to PM  staff (5 above), then from PM Staff to contractor (6 above), back to PM staff (7 above), then PM staff to finance 1 (8 above) 
- [x] from PM staff to contractor (11 above), back to PM staff (12 above), then from PM Staff to finance 1 (13 above), then from finance 1 to finance 2 (14 above).

Suggested Change:

- [x] when finance 2 requests changes the request goes from finance 2 to PM Staff (10 above) 
--> I disagree with the flow that finance 2 should go directly to PM. Missing step between finance 1 & 2. What if finance 2 has questions that are specifically for finance 1? Finance 1 approved a payment request, but finance 2 has more questions. finance 1 is the "maker" and finance 2 is the "checker" - these roles are not interchangeable 

Examples:
- Missing documents like an invoice, SAM check (finance 1 inserts a google link but the document was not in the folder), receipt, etc. 
- Document contains errors or inaccurate information. Finance 2 didn't see a period of performance/date on the invoice, the receipt was in a foreign language, dollar amount requested didn't match with receipt, etc.

- [x] contractor does not need a notification when PM sends invoice to finance 1 (4f above) 
- [x] does not need notifications for when invoices are sent to Intacct (15 above)

**Comment by ei8fdb** [comment]
> These are correct

Thanks for that, and also the other changes to the contractor notifications. 👍

> For example, what if finance 2 has questions or requests modifications that are specifically for finance 1?

Can you give me some more info on what those modifications could be? Just so I understand.

Are finance users to be able to modify an invoice submitted from a contractor?

**Comment by ei8fdb** [comment]
@fourthletter Thanks for the above comments Di. 

- [x] Can you confirm the invoice flow, and required notifications below are correct

![image](https://user-images.githubusercontent.com/1899506/152790379-df4dc61e-41ac-4397-9f17-9291fc9160a5.png)
(If it's hard to read, you [can see the original](https://notes.saneux.ei8fdb.org/shares/pI92hWbkbGPe09lysF9fWH))

Then we can:
- [x] start writing the content
- [ ] request sharing with PM and finance staff for feedback

**Comment by fourthletter** [comment]
@ei8fdb 
> Can you confirm the invoice flow, and required notifications below are correct

Workflow and required notifications look great -- thank you! During Feb 7 stand-up we decided default setting is staff gets slack notifications while contractors get emails.

**Comment by sks444** [comment]
@ei8fdb, thanks for the beautiful design!

Can you also add a notification flow for when Invoices gets declined?

All three users(PM Staff, Finance 1 and Finance 2) have the ability to decline invoices. I believe we are sending notification directly to contractor in all three cases?

cc: @fourthletter

**Comment by ei8fdb** [comment]
> @ei8fdb, thanks for the beautiful design!
> 
> Can you also add a notification flow for when Invoices gets declined?

Ah, you're 💯 right @sks444. We missed the case for invoice declining. Excellent catch. 

> All three users(PM Staff, Finance 1 and Finance 2) have the ability to decline invoices. I believe we are sending notification directly to contractor in all three cases?
> 
> cc: @fourthletter 

-  [x] That is how I understand it too. @fourthletter is that correct?

**Comment by fourthletter** [comment]
@ei8fdb 

> All three users(PM Staff, Finance 1 and Finance 2) have the ability to decline invoices. I believe we are sending notification directly to contractor in all three cases?

- [x] PM declines invoice - notifies Contractor

- [x] Finance 1 declines invoice - notifies PM; does NOT notify Contractor

- [x] Finance 2 declines invoice - notifies Finance 1; notifies PM? ; does NOT notify Contractor
- It's unclear to me if a decline from Finance 2 should notify PM because the issue may not be pertinent to the PM. For example, Finance 1 didn't complete a SAM check. Notifying the PM when they do not have agency to rectify the issue will only produce 'noise.'


Contractors do not need notifications until the PM has a grasp of the issue because:

Example: Invoice is for a 'partial' 20% of one deliverable. Finance 1 declined the invoice, and the PM needs to discuss this with either Finance OR the Contractor to resolve this issue. 

1. PM has a direct line of communication with the Contractor, PM's has the relationship and knows how to best communicate with the Contractor, e.g., via signal or a meeting. 
2. Not every scenario requires the Contractor's attention. A Contractor does not need to be notified if the PM and Finance 1 could resolve the issue internally.

**Comment by sks444** [comment]
Just to point to out that, in case where any changes are needed to an invoice we are using changes requested flow.

Declining an invoice makes the invoice invalid, that means it should not be processed further.

Invalid is something similar to deleting an invoice but still keeping the record in our system.

**Comment by sks444** [comment]
@ei8fdb, we continued this discussion on standup Feb 9.

We decided to only let PM/Staff to decline invoices.

PM has direct communication with both contractor and finance they can put up a valid reason to decline invoices. And also can communicate further with contractor.

There is very less possibility that an invoice get invalid once the PM has approved it. But even if that case occur finance can request changes. And PM can make it invalid.

**Comment by ei8fdb** [comment]
@sks444 

> There is very less possibility that an invoice get invalid once the PM has approved it. But even if that case occur finance can request changes. And PM can make it invalid.

I don't understand this part - can you explain it in more detail?

**Comment by sks444** [comment]
@ei8fdb, at finance level, we wouldn't generally want invoices to be declined when they have already been approved by PM.

That said, as finance we can always request changes back to PM. And if PM thinks those changes can not be made..they have the option to decline.

**Comment by ei8fdb** [comment]
> as finance we can always request changes back to PM. And if PM thinks those changes can not be made..they have the option to decline.

OK understood. So - there is no need for me to include this in the notification flow, or messages, correct?

**Comment by sks444** [comment]
>OK understood. So - there is no need for me to include this in the notification flow, or messages, correct?

Correct 💯

**Comment by ei8fdb** [comment]
The UX work is paused for the moment. I've asked for review from finance staff and devs. 

From discussion with Di yesterday, no design work is needed for the moment.

**Comment by fourthletter** [comment]
@sandeepsajan0 Could you please take a look at notifications for the finance workflow this month? We're planning to test with end-users, and it would be helpful for us to start drafting appropriate slack and email messages soon. Thanks.

**Comment by sandeepsajan0** [comment]
@fourthletter Yeah, sure. I picked this task and see that notifications are already implemented. So, I tested it properly for every user type (contractors, lead/staff, finance1, and finance2) and for all statuses (Submitted to Paid including decline) and its notifications work fine.

If you want to check or update those notification messages from one place(here) I can paste the notification message here. Let me know how you want to proceed.
